vm/loader/page_table/src/x64.rs

```diff
@@ -299,289 +392,311 @@ impl PteOps for PageTableBuilder {
-                            Self::build_pte(PageTableEntryType::Pde(output_address));
-                        self.set_pte_confidentiality(&mut new_entry, confidential);
-                        *pdpte_entry = new_entry;
-                        page_table.push(PageTable::new_zeroed());
-
-                        next_index
+            let mut get_or_insert_entry = |table_index: usize,
+                                           entry_level: EntryLevel,
+                                           current_va: &mut u64|
+             -> Option<usize> {
+                // If the current mapping can be inserted at this level as a leaf entry, do so
+                if (*current_va).is_multiple_of(entry_level.mapping_size())
+                    && (*current_va + entry_level.mapping_size() <= end_va)
+                {
+                    let entry = page_table[table_index].entry(*current_va, entry_level as u8);
+                    assert!(!entry.is_present());
+
+                    #[cfg(feature = "tracing")]
+                    tracing::trace!(
+                        "inserting entry for va: {:#X} at level {:?}",
+                        current_va,
+                        entry_level
+                    );
+
+                    let mut new_entry = PageTableBuilderParams::build_pte(
+                        entry_level.leaf(*current_va),
+                        permissions,
+                    );
+                    params.set_pte_confidentiality(&mut new_entry, confidential);
+                    *entry = new_entry;
+                    *current_va += entry_level.mapping_size() as u64;
+
+                    None
+                }
+                // The current mapping cannot be inserted as a leaf at this level
+                //
+                // Find or create the appropriate directory at this hierarchy level, and
+                // return the index
+                else {
+                    let directory_pa = entry_level.directory_pa(*current_va);
+                    let entry = page_table[table_index].entry(directory_pa, entry_level as u8);
+
+                    if !entry.is_present() {
+                        page_table_index += 1;
+                        // Allocate and link a page directory
+                        let output_address =
+                            params.page_table_gpa + page_table_index as u64 * X64_PAGE_SIZE;
+
+                        let mut new_entry = PageTableBuilderParams::build_pte(
+                            PageTableEntryType::Pde(output_address),
+                            permissions,
+                        );
+
+                        #[cfg(feature = "tracing")]
+                        tracing::trace!(
+                            "creating directory for va: {:#X} at level {:?}",
+                            directory_pa,
+                            entry_level
+                        );
+                        params.set_pte_confidentiality(&mut new_entry, confidential);
```

Comment:
it seems to me that either a build_pte_confidential would remove a lot of the `mut` usages in this code. Is it worth having `build_pte` and `build_pte_confidential`? 

In general, when i see a `mut` variable I try to think was this really needed to be mut or could the api be cleaned up to remove it?